enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,181 @@
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+If the `greenboot-default-health-check` RPM subpackage is installed, it already
+includes [health check scripts](https://github.com/fedora-iot/greenboot#health-checks-included-with-subpackage-greenboot-default-health-checks)
+verifying that DNS and `ostree` services can be accessed. 
+
+The MicroShift core RPM package will include a health check script to validate
+that all the required MicroShift services are up and running. The script will
+be installed into `/etc/greenboot/check/required.d` directory conditional to
+the presence of the `greenboot` package. This would prevent unnecessary scripts
+from being installed on systems that are not `rpm-ostree` based.
+
+The health check script will implement the following validations. Exiting the
+script with a non-zero status will have the boot declared as failed.
+
+|Validation                                          |Pass  |Fail  |
+|----------------------------------------------------|------|------|
+|microshift.service is enabled                       |Next  |exit 0|
+|Wait for any pods to start                          |Next  |exit 1|
+|Wait for `openshift-ovn-kubernetes` pods to be ready|Next  |exit 1|
+|Wait for `openshift-service-ca` pods to be ready    |Next  |exit 1|
+|Wait for `openshift-ingress` pods to be ready       |Next  |exit 1|
+|Wait for `openshift-dns` pods to be ready           |Next  |exit 1|
+|Wait for `openshift-storage` pods to be ready       |exit 0|exit 1|
+
+The wait period in each validation starts from 5 minutes and it is doubled
+after each boot in the verification loop. It should be possible to override the
+base wait period setting.
+
+### Risks and Mitigations
+
+Edge Devices often have relatively slow networks or occasional loss of network
+connectivity. The risk of declaring a `false positive` on the failed boot may
+be caused by these factors.
+
+If an update requires downloading new container images on a relatively slow
+network, it may take more than the default wait period to run this operation.
+The mitigation would be automatic doubling of the wait period on subsequent
+boots. If this is still not enough, system administrators should set the base
+wait period setting to a higher value to start with.
+
+If any external connectivity is required for upgrades to complete successfully,
+intermittent network outage may result in a rollback to the previous version.
+As a mitigation, system administrators should reinstall the upgrade when network
+is available for the Edge Devices.
```

Comment:
Another alternative is embedding the container images in the ostree image so that the new images will not need to be downloaded.

Reply:
or precaching on the device storage